chirp/drivers/tk11.py

```diff
@@ -0,0 +1,3297 @@
+        64286, 35833, 39896, 43963, 48026, 19061, 23124, 27191, 31254, 2801,
+        6864, 10931, 14994, 64814, 60687, 56684, 52557, 48554, 44427, 40424,
+        36297, 31782, 27655, 23652, 19525, 15522, 11395, 7392, 3265, 61215,
+        65342, 53085, 57212, 44955, 49082, 36825, 40952, 28183, 32310, 20053,
+        24180, 11923, 16050, 3793, 7920
+    ]
+
+    crc = 0
+    for value in content:
+        index = ((crc >> 8) ^ value) & 0xFF
+        crc = ((crc << 8) ^ num_array[index]) & 0xFFFF
+
+    return crc
+
+
+def content_nor_or(content: bytearray) -> None:
+    """content nor or"""
+    num_array = bytearray([
+        22, 108, 20, 230, 46, 145, 13, 64, 33, 53, 213, 64, 19, 3, 233, 128
+    ])
+
+    for i, byte in enumerate(content):
+        content[i] = byte ^ num_array[i % 16]
+
+
+def byte_nor_or(data: bytes, start_index: int, nor_or_len: int) -> bytes:
+    """byte nor or"""
+    content = bytearray(data[start_index:start_index + nor_or_len])
+
+    content_nor_or(content)
+
+    result = bytearray(data)
+    result[start_index:start_index + nor_or_len] = content
+
+    return bytes(result)
+
+
+def encapsulate_message(buf: bytes) -> bytes:
+    """Encapsulate a message with start and end message flags"""
+    length = len(buf)
+    header_bytes = struct.pack(
+        "<BBH",
+        START_MESSAGE_FLAG[0],
+        START_MESSAGE_FLAG[1],
+        length
+    )
+
+    crc = crc16_xmodem(buf)
+    ender_bytes = struct.pack(
+        "<HBB",
+        crc,
+        END_MESSAGE_FLAG[0],
+        END_MESSAGE_FLAG[1]
+    )
+
+    data = bytearray()
+    data.extend(header_bytes)
+    data.extend(buf)
+    data.extend(ender_bytes)
+
+    return byte_nor_or(data, 4, length + 2)
+
+
+def create_parameter_read_request(start_address: int, length: int):
+    """Create a parameter read request"""
+    address = 0x080000 + start_address
+
+    message_length = 0xC
+    rsv = (0, 0)
+
+    fmt = '<HHI H BB I'
+
+    return struct.pack(
+        fmt,
+        MessageType.MEMORY_READ_REQUEST.value,
+        message_length,
+        address,
+        length,
+        rsv[0],
+        rsv[1],
+        MAGIC_CODE
+    )
+
+
+def parse_data(serial_response) -> bytes | None:
+    """Parse data from serial response as bytes
+    None is returned if serial is empty or not valid"""
+    if serial_response is None or len(serial_response) < 4:
+        return None
+
+    length = serial_response[2] | (serial_response[3] << 8)
+    num_array1 = bytearray(length)
+    index = 4 + length + 2
+
+    if index > len(serial_response):
+        return None
+
+    if (serial_response[0] == 0xAB and
+            serial_response[1] == 0xCD and
+            serial_response[index] == 0xDC and
+            serial_response[index + 1] == 0xBA):
+        num_array1[:] = serial_response[4:4 + length]
+        return byte_nor_or(num_array1, 0, length)
+
+    return None
+
+
+def read_serial(serial, timeout_ms=2000) -> bytes | None:
```

Comment:
This is not an okay way to implement this. You should have a long (~1s) timeout and read the actual amount of data you're expecting. There is no reason to re-implement the serial line discipline from the OS. It's very inefficient but it's also very likely to not behave the same under all situations on all platforms. We used to have a bunch of drivers that worked like this and they ALL were constant trouble with people on various platforms, USB drivers/adapters, etc. I've mostly weeded out all of those out of the tree and it's been a massive stability improvement.

So please, read N bytes when you're expecting N. If you don't get N by the longer timeout, then it's an error.